bundles/org.openhab.binding.magentatv/src/main/java/org/openhab/binding/magentatv/internal/handler/MagentaTVHandler.java

```diff
@@ -0,0 +1,692 @@
+                if (account.isEmpty()) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "Credentials missing or invalid! Fill credentials into thing configuration or generate UID on the openHAB console - see README");
+                    return;
+                }
+
+                logger.info("{}: Authenticate account {}", thingId, account);
+                authenticateUser();
+            }
+
+            connectReceiver(); // throws MagentaTVException on error
+
+            // setup background device check
+            renewEventJob = scheduler.scheduleWithFixedDelay(new Runnable() {
+                @Override
+                public void run() {
+                    renewEventSubscription();
+                }
+            }, 2, 5, TimeUnit.MINUTES);
+
+            // change to ThingStatus.ONLINE will be done when the pairing result is received
+            // (see onPairingResult())
+        } catch (MagentaTVException e) {
+            errorMessage = e.toString();
+        } catch (RuntimeException e) {
+            logger.warn("{}: Exception on initialization", thingId, e);
+        } finally {
+            if (!errorMessage.isEmpty()) {
+                logger.debug("{}: {}", thingId, errorMessage);
+                setOnlineState(ThingStatus.OFFLINE, errorMessage);
+            }
+        }
+    }
+
+    /**
+     * This routine is called every time the Thing configuration has been changed (e.g. PaperUI)
+     */
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        logger.debug("{}: Thing config updated, re-initialize", thingId);
+        cancelAllJobs();
+        if (configurationParameters.containsKey(PROPERTY_ACCT_NAME)) {
+            @Nullable
+            String newAccount = (String) configurationParameters.get(PROPERTY_ACCT_NAME);
+            if ((newAccount != null) && !newAccount.isEmpty()) {
+                // new account info, need to renew userId
+                config.setUserID("");
+            }
+        }
+
+        super.handleConfigurationUpdate(configurationParameters);
+
+        // initializeThing();
+    }
+
+    /**
+     * Handle channel commands
+     *
+     * @param channelUID - the channel, which received the command
+     * @param command - the actual command (could be instance of StringType,
+     *            DecimalType or OnOffType)
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command == RefreshType.REFRESH) {
+            // currently no channels to be refreshed
+            return;
+        }
+
+        try {
+            if (!isOnline() || command.toString().equalsIgnoreCase("PAIR")) {
+                logger.debug("{}: Receiver {} is offline, try to (re-)connect", thingId, deviceName());
+                connectReceiver(); // reconnect to MR, throws an exception if this fails
```

Comment:
Can this be potentially long-running, e.g. when running in a timeout? The `handleCommand` method must not take more than 5 seconds to return.

Reply:
No, this is an asynchronous process, all good

Reply:
I am not convinced yet. The `connectReceiver()` method calls `subscribeEventChannel()` synchronously, which does a synchronous `http.sendData(...)`call.